Auto pruners

suiguoxin opened the pull request:
Add algo implementation / examples / test / doc for the following pruning algos:
- NetAdapt
- SimulatedAnnealing
- ADMM
- AutoCompress

chicm-ms commented on docs/en_US/Compressor/Pruner.md:
```diff
@@ -380,5 +385,168 @@ You can view example for more information
+- **sparsity:** How much percentage of convolutional filters are to be pruned.
+- **op_types:** Currently only Conv2d is supported in TaylorFOWeightFilterPruner.
+- **trainer:** Function used for the first step of ADMM training.
+- **evaluator:** Function to evaluate the masked model.
```
Suggest to describe the requirement, input/output of `trainer` and `evaluator`, and how they are used, better to provide an example.

suiguoxin replied:
Description added.

chicm-ms commented on src/sdk/pynni/nni/compression/torch/pruning/admm_pruner.py:
```diff
@@ -0,0 +1,201 @@
+        """
+
+        if self._pruning_mode == 'fine_grained':
+            return LevelPruner.calc_mask(self, wrapper, **kwargs)
```
L1FilterPruner can prune only Conv2d layer, should we check the configuration list and pruning mode together?

suiguoxin replied:
Fixed in validate_config.

chicm-ms commented on src/sdk/pynni/nni/compression/torch/pruning/admm_pruner.py:
```diff
@@ -0,0 +1,201 @@
+        if self._pruning_mode == 'fine_grained':
+            return LevelPruner.calc_mask(self, wrapper, **kwargs)
+        elif self._pruning_mode == 'channel':
+            return L1FilterPruner.calc_mask(self, wrapper, **kwargs)
```
`return L1FilterPruner.calc_mask(self, wrapper, **kwargs)` 
This is a nice way to integrate one Pruner into another, but it seems only work for those pruners which don't need intialization,  some pruners such as Slim,  Apoz , TalorOF , need initialization before being used.   Do you have requirement to use such pruners  in another Pruner?

chicm-ms replied:
The code directory structure is refactored, the `pruning` directory is planned to place the pure pruning algorithms which do not contain any training schedule. We can create another directory to place the Pruners which contains training schedule such as Anneal, Auto compress etc,   we can discuss it later.

suiguoxin replied:
Refactored with `Masker`.

chicm-ms commented on src/sdk/pynni/nni/compression/torch/pruning/net_adapt_pruner.py:
```diff
@@ -0,0 +1,410 @@
+        self._pruning_step = pruning_step
+
+        # overall pruning rate
+        self._sparsity = config_list[0]['sparsity']
```
We may need an overall configuration section in our config, especially for those algorithms with training schedule.

suiguoxin replied:
I prefer to refactor this in another PR, together with some related issues, e.g. the calc of overall sparsity.

chicm-ms commented on src/sdk/pynni/nni/compression/torch/pruning/net_adapt_pruner.py:
```diff
@@ -0,0 +1,410 @@
+        if count_non_prunable_ops:
+            pruner = LevelPruner(
+                copy.deepcopy(self._model_to_prune), [{'sparsity': 0.1, 'op_types': ['default']}])
+            modules_wrapper = pruner.get_modules_wrapper()
```
does the deepcopy support all models?

suiguoxin replied:
Yes, this is a solution recommended in the PyTorch forum to copy a model.

chicm-ms commented on src/sdk/pynni/nni/compression/torch/pruning/net_adapt_pruner.py:
```diff
@@ -0,0 +1,410 @@
+            for name, module in self.bound_model.named_modules():
+                if module == self.bound_model:
+                    continue
+                if self._pruning_mode == 'channel' and model_name in ['MobileNetV2', 'RetinaFace'] and name not in ops_no_dependency:
```
This checking seems very specific, would you explain the reason to do this check?

suiguoxin replied:
The PR https://github.com/microsoft/nni/pull/2435 provides functionalities to check dependencies. The checking here will be replaced after this PR is merged.

suiguoxin replied:
The checking is removed. To make it clear, I'll deal with this functionality (prune only ops without dependency) in a seperate PR in the future.

suiguoxin commented:
Changes since last review meeting:
- Resolve confict after # PR https://github.com/microsoft/nni/pull/2504
- Refactor ADMMPruner with `Masker` : not convenient to use `Masker` in other auto-pruners (ADMMPruner is not an auto-pruner, it does not deal with automatic sparsity distribution)
- Rename of some common class arguments:
    - Pruning_mode (channel, fine_grained)-> base_algo (level, l1, l2)
    - Fine_tuner -> short_term_fine_tune 
    -  Pruning_step -> sparsity_per_iteration (NetAdaptPruner)
    - other specific args

- Detailed explanation of function (args, return value):
    - evaluator
    - short_term_fine_tuner
    - trainer (ADMMPruner, AutoCompressPruner)
- Include l2 as an option of base_pruning_algo
- Remove check for specific models in pruner implementation
- NetAdaptPruner:
    - Function Check related ops: Fix known issues for some cases: 'dropout'/'pooling'… etc


Not done:
- Refactor auto-pruners(SAPruner/NetAdaptPruner) with `Masker`:
    - Not convenient to use `Masker` class to generate masks for all the ops at once given the config_list
-  Remove `speedup` from AutoCompressPruner:
    - This will involves huge changes to SimulatedAnnealing & ADMMPruner and make them difficult to maintain. As the `speedup` component is always necessary for application and we are supporting more and more models in speedup, I suggest to keep it for the moment.

chicm-ms commented on src/sdk/pynni/nni/compression/torch/pruning/admm_pruner.py:
```diff
@@ -0,0 +1,197 @@
+        dict
+            dictionary for storing masks
+        """
+        self.masker = MASKER_DICT[self._base_algo](self.bound_model, self)
```
This `self.masker` is not used.

suiguoxin replied:
> This `self.masker` is not used.
`self.masker` is required by OneShotPruner to realize `calc_mask`

suiguoxin replied:
Moved to `__init__` .

QuanluZhang commented on docs/en_US/Compressor/Pruner.md:
```diff
@@ -390,5 +395,170 @@ You can view example for more information
 
+## NetAdapt Pruner
+NetAdapt allows a user to automatically simplify a pretrained network to meet the resource budget. 
+For example, given the ovrall sparsity, NetAdapt will automatically generate the sparsities distribution among different layers by iterative pruning.
```
ovrall -> overall

suiguoxin replied:
Fixed

QuanluZhang commented on docs/en_US/Compressor/Pruner.md:
```diff
@@ -390,5 +395,170 @@ You can view example for more information
+#### User configuration for NetAdapt Pruner
+
+- **sparsity:** How much percentage of convolutional filters are to be pruned.
+- **op_types:** Currently only Conv2d is supported in TaylorFOWeightFilterPruner.
```
in TaylorFOWei...?

suiguoxin replied:
Fixed

QuanluZhang commented on docs/en_US/Compressor/Pruner.md:
```diff
@@ -390,5 +395,170 @@ You can view example for more information
+This function should include `model` as the only parameter, and fine tune the model for a short term after each pruning iteration.
+- **evaluator:** Function to evaluate the masked model. This function should include `model` as the only parameter, and returns a scalar value.
+- **base_algo:** base pruning algorithm. 'level', 'l1' or 'l2', by default 'l1'.
+- **sparsity_per_iteration:** sparsity to prune in each iteration
```
better to give more explanation on this argument. give the final sparsity 0.8. if sparsity_per_iteration is 01. Do you mean the total sparsity would be 0, 0.1, 0.2, 0.3, ..., 0.8?

suiguoxin replied:
Yes. More explanation added.

QuanluZhang commented on docs/en_US/Compressor/Pruner.md:
```diff
@@ -390,5 +395,170 @@ You can view example for more information
+- **op_types:** Currently only Conv2d is supported in ADMMPruner.
+- **trainer:** Function used for the first optimization subproblem.
+This function should include `model, optimizer, criterion, epoch, callback` as parameters, where callback should be inserted after loss.backward of the normal training process.
+- **optimize_iteration:** ADMM optimize iterations.
```
you mean the number of iterations?

suiguoxin replied:
Yes and renamed.

QuanluZhang commented on docs/en_US/Compressor/Pruner.md:
```diff
@@ -390,5 +395,170 @@ You can view example for more information
+- **trainer:** Function used for the first optimization subproblem.
+This function should include `model, optimizer, criterion, epoch, callback` as parameters, where callback should be inserted after loss.backward of the normal training process.
+- **optimize_iteration:** ADMM optimize iterations.
+- **training_epochs:** training epochs of the first optimization subproblem.
```
it is not clear what is "the first optimization subproblem". better to give a little more description in the introduction of this pruner

suiguoxin replied:
Added.

QuanluZhang commented on docs/en_US/Compressor/Pruner.md:
```diff
@@ -390,5 +395,170 @@ You can view example for more information
+- **optimize_iteration:** ADMM optimize iterations.
+- **training_epochs:** training epochs of the first optimization subproblem.
+- **row:** penalty parameters for ADMM training.
+- **base_algo:** base pruning algorithm. 'level', 'l1' or 'l2', by default 'l1'.
```
why this one does not have `experiment_data_dir`?

suiguoxin replied:
ADMMPruner is not an auto pruner, there is thus no experiment data generated. More explanation on what is included as experiment data added for the auto pruners.

QuanluZhang commented on docs/en_US/Compressor/Pruner.md:
```diff
@@ -390,5 +395,170 @@ You can view example for more information
+            
+
+## AutoCompress Pruner
+For each round t, AutoCompressPruner prune the model for the same sparsity each round to achive the ovrall sparsity:
```
ovrall -> overall

suiguoxin replied:
Fixed

QuanluZhang commented on docs/en_US/Compressor/Pruner.md:
```diff
@@ -390,5 +395,170 @@ You can view example for more information
+
+- **sparsity:** How much percentage of convolutional filters are to be pruned.
+- **op_types:** "Conv2d" or "default".
+- **trainer:** Function used for the first optimization subproblem.
```
it is not clear how to write `trainer`. who should provide `callback`? what is the reason to provide `callback`? why it should be put behind `loss.backward`?

suiguoxin replied:
updated

QuanluZhang commented on docs/en_US/Compressor/Pruner.md:
```diff
@@ -390,5 +395,170 @@ You can view example for more information
+- **trainer:** Function used for the first optimization subproblem.
+This function should include `model, optimizer, criterion, epoch, callback` as parameters, where callback should be inserted after loss.backward of the normal training process.
+- **evaluator:** Function to evaluate the masked model. This function should include `model` as the only parameter, and returns a scalar value.
+- **dummy_input:** The dummy input for model speed up, users should put it on right device before pass in.
```
why there is model speed up here?

suiguoxin replied:
`speedup` is called inside `AutoCompress` to keep the model un-masked and realize real pruning after each iteration.

QuanluZhang commented on docs/en_US/Compressor/Pruner.md:
```diff
@@ -390,5 +395,170 @@ You can view example for more information
+- **evaluator:** Function to evaluate the masked model. This function should include `model` as the only parameter, and returns a scalar value.
+- **dummy_input:** The dummy input for model speed up, users should put it on right device before pass in.
+- **iterations:** The number of overall iterations.
+- **optimize_mode:** Optimize mode, 'maximize' or 'minimize', by default 'maximize'.
```
only this auto pruner supports `optimize_mode`?

suiguoxin replied:
`optimize_mode` is supported in `NetAdapt` `SimualatedAnnealing` and `AucoCompress`. Sorry for have missed this arg for `NetAdaptPruner`.

QuanluZhang commented on docs/en_US/Compressor/Pruner.md:
```diff
@@ -390,5 +395,170 @@ You can view example for more information
+- **stop_temperature:** Simualated Annealing related parameter.
+- **cool_down_rate:** Simualated Annealing related parameter.
+- **perturbation_magnitude:** Initial perturbation magnitude to the sparsities. The magnitude decreases with current temperature.
+- **optimize_iteration:** ADMM optimize iterations.
```
what is the relation with ADMM?

suiguoxin replied:
AutoCompress Pruner call `SimualtedAnnealing Pruner`  and `ADMM Pruner ` iteratively.

QuanluZhang commented on docs/en_US/Compressor/Pruner.md:
```diff
@@ -390,5 +395,170 @@ You can view example for more information
+- **cool_down_rate:** Simualated Annealing related parameter.
+- **perturbation_magnitude:** Initial perturbation magnitude to the sparsities. The magnitude decreases with current temperature.
+- **optimize_iteration:** ADMM optimize iterations.
+- **epochs:** training epochs of the first optimization subproblem.
```
this one also has two subproblems?

suiguoxin replied:
These are args for `ADMM`

suiguoxin commented on src/sdk/pynni/nni/compression/torch/pruning/net_adapt_pruner.py:
```diff
@@ -0,0 +1,324 @@
+            and fine tune the model for a short term after each pruning iteration.
+        optimize_mode : str
+            optimize mode, 'maximize' or 'minimize', by default 'maximize'
+        base_algo : str
```
added

suiguoxin commented on src/sdk/pynni/nni/compression/torch/pruning/net_adapt_pruner.py:
```diff
@@ -0,0 +1,324 @@
+        1. Con = Res_i - delta_Res
+        2. for every layer:
+            Choose Num Filters to prune
+            Choose which filter to prunee
```
fixed

suiguoxin commented on src/sdk/pynni/nni/compression/torch/pruning/auto_compress_pruner.py:
```diff
@@ -0,0 +1,221 @@
+        """
+        _logger.info('Starting AutoCompress pruning...')
+
+        sparsity_each_round = 1 - pow(1-self._sparsity, 1/self._optimize_iterations)
```
This strategy is used to ensure that same number of weights will be pruned in each iteration.

chicm-ms commented on docs/en_US/Compressor/Pruner.md:
```diff
@@ -398,5 +402,176 @@ We try to reproduce the experiment result of the fully connected network on MNIS
 The above figure shows the result of the fully connected network. `round0-sparsity-0.0` is the performance without pruning. Consistent with the paper, pruning around 80% also obtain similar performance compared to non-pruning, and converges a little faster. If pruning too much, e.g., larger than 94%, the accuracy becomes lower and convergence becomes a little slower. A little different from the p ...
 
 
+## NetAdapt Pruner
```
The order of each section is better consistent with the content directory at beginning.

suiguoxin replied:
fixed

colorjam commented on src/sdk/pynni/nni/compression/torch/pruning/auto_compress_pruner.py:
```diff
@@ -0,0 +1,227 @@
+
+            # use speed up to prune the model before next iteration, because SimulatedAnnealingPruner & ADMMPruner don't take masked models
+            self._model_to_prune.load_state_dict(torch.load(os.path.join(
+                self._experiment_data_dir, 'model_admm_masked.pth')))
```
Why reload the checkpoint?

suiguoxin replied:
The model weights have changed after admm pruner.

colorjam commented on src/sdk/pynni/nni/compression/torch/pruning/admm_pruner.py:
```diff
@@ -0,0 +1,181 @@
+            Penalty parameters for ADMM training.
+        base_algo : str
+            Base pruning algorithm. `level`, `l1` or `l2`, by default `l1`.
+            Given the sparsity distrution among the ops, the assigned `base_algo` is used to decide which filters/channels/weights to prune.
```
distrution -> distribution

suiguoxin replied:
fixed
